File: ruby/10-null-spiral/src/room_lookup.rb
# frozen_string_literal: true

class RoomLookup
  def get_room_for_student(student_id, db)
    return nil if db.nil?

    student = db.get_student(student_id)
    return nil if student.nil?

    enrollment = db.get_enrollment(student[:id])
    return nil if enrollment.nil?

    course = db.get_course(enrollment[:course_id])
    return nil if course.nil?

    section = db.get_section(course[:default_section_id])
    return nil if section.nil?

    room = db.get_room(section[:room_id])
    return room[:name] unless room.nil?

    nil
  end
end

File: javascript/10-null-spiral/src/room_lookup.js
class RoomLookup {
  constructor() {}

  get_room_for_student(studentId, db) {
    if (db !== null) {
      let student = db.getStudent(studentId);
      if (student !== null) {
        let enrollment = db.getEnrollment(student.id);
        if (enrollment !== null) {
          let course = db.getCourse(enrollment.courseId);
          if (course !== null) {
            let section = db.getSection(course.defaultSectionId);
            if (section !== null) {
              let room = db.getRoom(section.roomId);
              if (room !== null) {
                return room.name;
              } else {
                return null;
              }
            } else {
              return null;
            }
          } else {
            return null;
          }
        } else {
          return null;
        }
      } else {
        return null;
      }
    } else {
      return null;
    }
  }
}

export { RoomLookup };
